File: scripts/inspect_processdata.py
#!/usr/bin/env python3
"""
Inspect processdata.h5 using Python h5py to understand its true structure
"""

import h5py
import numpy as np

def inspect_dataset(name, obj):
    """Inspect a dataset and print its details"""
    if isinstance(obj, h5py.Dataset):
        print(f"\n{'='*80}")
        print(f"📊 Dataset: {name}")
        print(f"{'='*80}")
        print(f"  Shape: {obj.shape}")
        print(f"  Dtype: {obj.dtype}")
        print(f"  Size: {obj.size} elements")
        print(f"  Chunks: {obj.chunks}")
        print(f"  Compression: {obj.compression}")
        print(f"  Compression opts: {obj.compression_opts}")
        
        # Check if it's a virtual dataset
        if hasattr(obj, 'is_virtual') and obj.is_virtual:
            print(f"  ⚠️  VIRTUAL DATASET")
            try:
                print(f"  Virtual sources: {obj.virtual_sources()}")
            except:
                print(f"  (Cannot get virtual sources)")
        
        # Print attributes
        if len(obj.attrs) > 0:
            print(f"\n  🏷️  Attributes ({len(obj.attrs)}):")
            for attr_name, attr_value in obj.attrs.items():
                print(f"     - {attr_name}: {attr_value}")
        
        # Try to read data
        print(f"\n  📖 Attempting to read data...")
        try:
            data = obj[...]
            print(f"  ✅ Success! Shape: {data.shape}, Dtype: {data.dtype}")
            
            if data.size <= 10:
                print(f"  Data: {data}")
            else:
                print(f"  First 5 elements: {data.flat[:5]}")
                print(f"  Last 5 elements: {data.flat[-5:]}")
                print(f"  Min: {np.min(data)}, Max: {np.max(data)}, Mean: {np.mean(data)}")
        except Exception as e:
            print(f"  ❌ Failed to read: {e}")

def inspect_group(name, obj):
    """Inspect a group and print its details"""
    if isinstance(obj, h5py.Group):
        print(f"\n{'='*80}")
        print(f"📁 Group: {name}")
        print(f"{'='*80}")
        print(f"  Children: {list(obj.keys())}")
        
        # Print attributes
        if len(obj.attrs) > 0:
            print(f"\n  🏷️  Attributes ({len(obj.attrs)}):")
            for attr_name, attr_value in obj.attrs.items():
                print(f"     - {attr_name}: {attr_value}")

def main():
    filename = 'example/data/processdata.h5'
    
    print("="*80)
    print(f"🔬 Python h5py Inspection: {filename}")
    print("="*80)
    
    try:
        with h5py.File(filename, 'r') as f:
            print(f"\n📋 File Information:")
            print(f"   HDF5 Version: {h5py.version.hdf5_version}")
            print(f"   h5py Version: {h5py.version.version}")
            print(f"   Root keys: {list(f.keys())}")
            
            # Inspect root attributes
            if len(f.attrs) > 0:
                print(f"\n🏷️  Root Attributes ({len(f.attrs)}):")
                for attr_name, attr_value in f.attrs.items():
                    print(f"   - {attr_name}: {attr_value}")
            
            # Visit all objects
            print(f"\n{'='*80}")
            print("📂 Visiting all objects...")
            print(f"{'='*80}")
            
            def visitor(name, obj):
                if isinstance(obj, h5py.Dataset):
                    inspect_dataset(name, obj)
                elif isinstance(obj, h5py.Group):
                    inspect_group(name, obj)
            
            f.visititems(visitor)
            
            print(f"\n{'='*80}")
            print("✅ Inspection complete")
            print(f"{'='*80}")
            
    except Exception as e:
        print(f"\n❌ Error opening file: {e}")
        import traceback
        traceback.print_exc()

if __name__ == '__main__':
    main()

File: scripts/inspect_processdata.dart
import 'package:dartframe/dartframe.dart';

Future<void> inspectDataset(Hdf5File file, String path,
    {bool detailed = false}) async {
  print('\n📊 Dataset: $path');

  try {
    final ds = await file.dataset(path);
    print('  ✓ Successfully opened');
    print('  📐 Shape: ${ds.shape}');
    print('  🔤 Datatype: ${ds.datatype}');
    print('  📦 Layout: ${ds.layout.runtimeType}');

    if (ds.layout is ChunkedLayout) {
      final chunked = ds.layout as ChunkedLayout;
      print('  🧩 Chunk dimensions: ${chunked.chunkDimensions}');
      print('  📍 Chunk address: 0x${chunked.address.toRadixString(16)}');
    } else if (ds.layout is ContiguousLayout) {
      final contiguous = ds.layout as ContiguousLayout;
      print('  📍 Data address: 0x${contiguous.address.toRadixString(16)}');
      print('  📏 Data size: ${contiguous.size} bytes');
    }

    if (ds.filterPipeline != null && ds.filterPipeline!.isNotEmpty) {
      print('  🔧 Filters: ${ds.filterPipeline}');
    }

    // Show attributes
    final attrs = ds.attributes;
    if (attrs.isNotEmpty) {
      print('  🏷️  Attributes (${attrs.length}):');
      for (final attr in attrs) {
        print('     - ${attr.name}: ${attr.value}');
      }
    }

    // Try to read data
    if (detailed) {
      try {
        print('  📖 Attempting to read data...');
        final data = await file.readDataset(path);
        print('  ✅ Read successful: ${data.length} elements');

        if (data.isNotEmpty) {
          if (data.length <= 5) {
            print('  📄 Data: $data');
          } else {
            print('  📄 First 3 elements: ${data.take(3).toList()}');
            print(
                '  📄 Last 3 elements: ${data.skip(data.length - 3).toList()}');
          }
        }
      } catch (e) {
        print('  ❌ Read failed: ${e.toString().split('\n').first}');
      }
    }
  } catch (e) {
    print('  ❌ Failed to open: ${e.toString().split('\n').first}');

    // Try to get more details by reading the object header directly
    try {
      final group = await file.group('/');
      final address = group.getChildAddress(path.substring(1));

      if (address != null) {
        print('  🔍 Object address: 0x${address.toRadixString(16)}');

        final fileIO = FileIO();
        final raf =
            await fileIO.openRandomAccess('example/data/processdata.h5');
        final reader = ByteReader(raf);

        try {
          final header = await ObjectHeader.read(reader, address,
              filePath: 'example/data/processdata.h5');

          print('  📋 Object header messages: ${header.messages.length}');

          final datatype = header.findDatatype();
          if (datatype != null) {
            print('  🔤 Datatype found: $datatype');
          }

          final dataspace = header.findDataspace();
          if (dataspace != null) {
            print('  📐 Dataspace found: ${dataspace.dimensions}');
          }

          final layout = header.findDataLayout();
          if (layout != null) {
            print('  📦 Layout found: ${layout.runtimeType}');
          }
        } finally {
          await raf.close();
        }
      }
    } catch (e2) {
      print(
          '  🔍 Additional details unavailable: ${e2.toString().split('\n').first}');
    }
  }
}

Future<void> inspectObject(Hdf5File file, String path) async {
  try {
    final objType = await file.getObjectType(path);

    if (objType == 'dataset') {
      await inspectDataset(file, path, detailed: true);
    } else if (objType == 'group') {
      print('\n📁 Group: $path');
      final group = await file.group(path);
      print('  Children: ${group.children.join(", ")}');

      // Show attributes
      final attrs = group.header.findAttributes();
      if (attrs.isNotEmpty) {
        print('  🏷️  Attributes (${attrs.length}):');
        for (final attr in attrs) {
          print('     - ${attr.name}: ${attr.value}');
        }
      }
    } else {
      print('\n❓ Unknown object: $path');
      print('  Type: $objType');
    }
  } catch (e) {
    print('\n❌ Error accessing $path: ${e.toString().split('\n').first}');
  }
}

void main() async {
  print('═' * 80);
  print('🔬 Detailed Inspection: example/data/processdata.h5');
  print('═' * 80);

  try {
    final file = await Hdf5File.open('example/data/processdata.h5');

    try {
      // Get root children
      final rootChildren = file.list('/');
      print('\n📁 Root level objects: ${rootChildren.length}');
      print('   ${rootChildren.join(", ")}');

      // Inspect each object
      for (final child in rootChildren) {
        await inspectObject(file, '/$child');
      }

      print('\n${'═' * 80}');
      print('✅ Inspection complete');
      print('═' * 80);
    } finally {
      await file.close();
    }
  } catch (e) {
    print('\n❌ Error opening file: $e');
  }
}
